homeassistant/components/cloudflare/__init__.py

```diff
@@ -97,22 +93,46 @@ async def _async_update_cloudflare(
 ) -> None:
     _LOGGER.debug("Starting update for zone %s", dns_zone["name"])
 
-    records = await client.list_dns_records(zone_id=dns_zone["id"], type="A")
-    _LOGGER.debug("Records: %s", records)
-
-    session = async_get_clientsession(hass, family=socket.AF_INET)
-    location_info = await async_detect_location_info(session)
-
-    if not location_info or not is_ipv4_address(location_info.ip):
-        raise HomeAssistantError("Could not get external IPv4 address")
-
-    filtered_records = [
-        record
-        for record in records
-        if record["name"] in target_records and record["content"] != location_info.ip
-    ]
-
-    if len(filtered_records) == 0:
+    old_records = await client.list_dns_records(zone_id=dns_zone["id"])
+    _LOGGER.debug("Records: %s", old_records)
+
+    records_to_be_updated = []
+
+    for record in old_records:
+        if record["name"] not in target_records:
+            continue
+        if record["type"] == "A":
+            _LOGGER.debug("Found target record: %s", record["name"])
+            session = async_get_clientsession(hass, family=socket.AF_INET)
+            location_info = await async_detect_location_info(session)
+            if location_info and is_ipv4_address(location_info.ip):
+                _LOGGER.debug("IPv4 address detected: %s", location_info.ip)
+                if record["content"] != location_info.ip:
+                    _LOGGER.info(
+                        "IPv4 address change detected for record: %s,will DNS entry from %s to %s",
+                        record["name"],
+                        record["content"],
+                        location_info.ip,
+                    )
+                    record["content"] = location_info.ip
+                    records_to_be_updated.append(record)
+        if record["type"] == "AAAA":
+            _LOGGER.debug("Found target record: %s", record["name"])
+            session_ipv6 = async_get_clientsession(hass, family=socket.AF_INET6)
+            location_info_v6 = await async_detect_location_info(session_ipv6)
+            if location_info_v6 and is_ipv6_address(location_info_v6.ip):
+                _LOGGER.debug("IPv6 address detected: %s", location_info_v6.ip)
+                if record["content"] != location_info_v6.ip:
+                    _LOGGER.info(
+                        "IPv6 address change detected for record: %s,will DNS entry from %s to %s",
+                        record["name"],
+                        record["content"],
+                        location_info_v6.ip,
+                    )
+                    record["content"] = location_info_v6.ip
+                    records_to_be_updated.append(record)
```

Comment:
The only difference between these 2 is the client session right? Can we maybe deduplicate this code more so we only set the right client session and just go on from there?

Reply:
https://github.com/home-assistant/core/pull/129809#discussion_r1863197639 would be better IMO.
(will also remove the duplicated parts.

Reply:
I'm not sure what comment exactly you're taking about

Reply:
> Instead of looping like this, extract the correct record form the list, and act on that object (if found).